{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3af8b94a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import scipy.stats\n",
    "from decimal import Decimal "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a42fdad",
   "metadata": {},
   "source": [
    "\n",
    "Deleteriousness burden, 417 CRISPR hits with 10+ variants\n",
    "\n",
    "Regeneron burden, up to 490 CRISPR hits (may be fewer depending on which genes have variants)\n",
    "\n",
    "Deleteriousness burden, genome-wide\n",
    "\n",
    "Regeneron burden, genome-wide\n",
    "\n",
    "Deleteriousness burden, module 255 (up to 14 genes, may be fewer depending on which have variants)\n",
    "\n",
    "Regeneron burden, module 255 (up to 14 genes, may be fewer depending on which have variants)\n",
    "\n",
    "Deleteriousness burden, module 69 (54 genes, may be fewer depending on which have variants)\n",
    "\n",
    "Regeneron burden, module 69 (54 genes, may be fewer depending on which have variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fe05693",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_harmoic_p(p_list, weights = None):\n",
    "    if weights is None:\n",
    "        weights = [1 / len(p_list)] * len(p_list)\n",
    "    numerator = np.sum(weights)\n",
    "    denominator = 0\n",
    "    for w,p in zip(weights, p_list):\n",
    "        denominator += (w/p)\n",
    "    return numerator / denominator \n",
    "\n",
    "def simes(p_vals):\n",
    "    sorted_p = p_vals.sort_values()\n",
    "    ranks = pd.Series(range(1, len(p_vals) + 1))\n",
    "    multiplied = sorted_p * len(ranks)\n",
    "    results = multiplied/ranks.values\n",
    "    return min(results)\n",
    "\n",
    "def get_browns_chi_sq(p_list, dof):\n",
    "    summation = 0\n",
    "    for p in p_list:\n",
    "        summation += np.exp(p)\n",
    "    X = -2 * summation \n",
    "    print(X)\n",
    "    return scipy.stats.chi2.cdf(X, dof)\n",
    "\n",
    "def format_p(p):\n",
    "    if p < 0.0001:\n",
    "        sci_notation_pieces = \"{:.2e}\".format(p).split(\"e\")\n",
    "        exponent = int(sci_notation_pieces[1])\n",
    "        decimal = sci_notation_pieces[0]\n",
    "        p_string =  decimal + \" x 10^{\" + str(exponent) + \"}\" +  \"$\"\n",
    "        return p_string\n",
    "    else:\n",
    "        return round(p , 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c16c849",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"490_genes_from_cell_screen.txt\", \"r\") as f:\n",
    "    genes_490 = list(map(lambda x : x.replace(\"\\n\", \"\"), f.readlines()))\n",
    "f.close()\n",
    "\n",
    "with open(\"LDL_inc_genes_from_screen.txt\", \"r\") as f:\n",
    "    genes_increase = list(map(lambda x : x.replace(\"\\n\", \"\"), f.readlines()))\n",
    "f.close()\n",
    "\n",
    "with open(\"LDL_dec_genes_from_screen.txt\", \"r\") as f:\n",
    "    genes_decrease = list(map(lambda x : x.replace(\"\\n\", \"\"), f.readlines()))\n",
    "f.close()\n",
    "\n",
    "df = pd.read_csv(\"./GSEA_summaries/perc_25_d_2_100k_iterations.csv\")\n",
    "\n",
    "module_69_genes = df.loc[df[\"Term\"] == \"module_69\"].iloc[0][\"genes\"].split(\";\")\n",
    "module_255_genes = df.loc[df[\"Term\"] == \"module_255\"].iloc[0][\"genes\"].split(\";\")\n",
    "\n",
    "all_vals = []"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f53ffb45",
   "metadata": {},
   "source": [
    "### Deleteriousness burden, 417 CRISPR hits with 10+ variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "687f07b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mega_df = pd.concat([\n",
    "    pd.read_csv('./alternate_score_files/042722_whole_genome_0th_perc_442_genes.csv'),\n",
    "    pd.read_csv(f\"041522_whole_genome_25th_perc.csv\"),\n",
    "    pd.read_csv(f\"041522_whole_genome_50th_perc.csv\"),\n",
    "    pd.read_csv(f\"041522_whole_genome_75th_perc.csv\")\n",
    "])\n",
    "\n",
    "df = mega_df\n",
    "\n",
    "print(len(df))\n",
    "\n",
    "df[\"mwu_a_pval\"] = df[['mwu_a_pval_greater', 'mwu_a_pval_less']].min(axis=1)\n",
    "df[\"mwu_two_sided\"] = df[\"mwu_a_pval\"].apply(lambda x : x*2)\n",
    "df[\"mwu_a_pval_intra\"] = df[['mwu_a_pval_intra_greater','mwu_a_pval_intra_less']].min(axis=1)\n",
    "\n",
    "col_of_interest = \"mwu_a_pval\"\n",
    "\n",
    "subset = df.loc[\n",
    "    (df[\"gene\"].isin(genes_490)) &\n",
    "    (~df[col_of_interest].isna()) &\n",
    "    (df['high_impact_carriers'] >= 10)\n",
    "].sort_values(by = col_of_interest)\n",
    "\n",
    "alt_values = []\n",
    "for g in set(subset[\"gene\"].values):\n",
    "    s = subset.loc[subset[\"gene\"] == g]\n",
    "    p_vals = pd.Series(s[col_of_interest].values)\n",
    "    simes_p = simes(p_vals)\n",
    "    harmonic_p = get_harmoic_p(p_vals.values)\n",
    "    alt_values.append([g, simes_p, harmonic_p])\n",
    "    \n",
    "simes_df = pd.DataFrame(alt_values, columns = [\"gene\", \"simes_p\", \"harmonic_p\"])\n",
    "m = len(simes_df)\n",
    "print(m)\n",
    "print(\"------\")\n",
    "\n",
    "signif = set()\n",
    "alpha = 0.1\n",
    "simes_df = simes_df.sort_values(by = \"simes_p\")\n",
    "for counter, (index, row) in enumerate(simes_df.iterrows()):\n",
    "    if row[\"simes_p\"] <= ((counter+1) / m) * alpha:\n",
    "        signif.add(row[\"gene\"])\n",
    "    \n",
    "print(len(signif))\n",
    "print(\", \".join(list(signif)))\n",
    "\n",
    "title = \"Deleteriousness burden, 417 CRISPR hits with 10+ variant\"\n",
    "all_vals.append([title, m, len(signif), \"|\".join(sorted(list(signif)))])\n",
    "print(\"less than 0.05\", len(simes_df.loc[simes_df[\"simes_p\"] <= 0.05]))\n",
    "print( \"|\".join(sorted(list(signif))))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27ff5fd2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "af0ceaa6",
   "metadata": {},
   "source": [
    "###  Regeneron burden, up to 490 CRISPR hits (may be fewer depending on which genes have variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42e171d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e78b91e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"./data/GCST90083019_buildGRCh38.tsv\", delimiter = \"\\t\")\n",
    "df[\"gene\"] = df[\"Name\"].apply(lambda x : x.split(\"(\")[0])\n",
    "df = df.loc[\n",
    "    (df[\"gene\"].isin(genes_490)) &\n",
    "    (~df[\"p_value\"].isna())\n",
    "]\n",
    "alt_values = []\n",
    "for g in set(df[\"gene\"].values):\n",
    "    s = df.loc[df[\"gene\"] == g]\n",
    "    p_vals = pd.Series(s[\"p_value\"].values)\n",
    "    simes_p = simes(p_vals)\n",
    "    harmonic_p = get_harmoic_p(p_vals.values)\n",
    "    alt_values.append([g, simes_p, harmonic_p])\n",
    "    \n",
    "simes_df = pd.DataFrame(alt_values, columns = [\"gene\", \"simes_p\", \"harmonic_p\"])\n",
    "\n",
    "signif = set()\n",
    "alpha = 0.1\n",
    "simes_df = simes_df.sort_values(by = \"simes_p\")\n",
    "m = len(simes_df)\n",
    "\n",
    "for counter, (index, row) in enumerate(simes_df.iterrows()):\n",
    "    if row[\"simes_p\"] <= ((counter+1) / m) * alpha:\n",
    "        signif.add(row[\"gene\"])\n",
    "\n",
    "print(len(signif), m)\n",
    "print(\", \".join(sorted(list(signif))))\n",
    "print(\"less than 0.05\", len(simes_df.loc[simes_df[\"simes_p\"] <= 0.05]))\n",
    "\n",
    "title = \"Regeneron burden, up to 490 CRISPR hits (may be fewer depending on which genes have variants)\"\n",
    "all_vals.append([title, m, len(signif), \"|\".join(sorted(list(signif)))])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b74a48c3",
   "metadata": {},
   "source": [
    "### Deleteriousness burden, genome-wide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bdc9419",
   "metadata": {},
   "outputs": [],
   "source": [
    "mega_df = pd.concat([\n",
    "    pd.read_csv('041522_whole_genome_0th_perc.csv'),\n",
    "    pd.read_csv(f\"041522_whole_genome_25th_perc.csv\"),\n",
    "    pd.read_csv(f\"041522_whole_genome_50th_perc.csv\"),\n",
    "    pd.read_csv(f\"041522_whole_genome_75th_perc.csv\")\n",
    "])\n",
    "\n",
    "df = mega_df\n",
    "\n",
    "print(len(df))\n",
    "\n",
    "df[\"mwu_a_pval\"] = df[['mwu_a_pval_greater', 'mwu_a_pval_less']].min(axis=1)\n",
    "df[\"mwu_two_sided\"] = df[\"mwu_a_pval\"].apply(lambda x : x*2)\n",
    "df[\"mwu_a_pval_intra\"] = df[['mwu_a_pval_intra_greater','mwu_a_pval_intra_less']].min(axis=1)\n",
    "\n",
    "\n",
    "col_of_interest = \"mwu_a_pval\"\n",
    "\n",
    "subset = df.loc[\n",
    "#     (df[\"gene\"].isin(genes_490)) &\n",
    "    (~df[col_of_interest].isna()) &\n",
    "    (df['high_impact_carriers'] >= 10)\n",
    "].sort_values(by = col_of_interest)\n",
    "\n",
    "alt_values = []\n",
    "for g in set(subset[\"gene\"].values):\n",
    "    s = subset.loc[subset[\"gene\"] == g]\n",
    "    p_vals = pd.Series(s[col_of_interest].values)\n",
    "    simes_p = simes(p_vals)\n",
    "    harmonic_p = get_harmoic_p(p_vals.values)\n",
    "    alt_values.append([g, simes_p, harmonic_p])\n",
    "    \n",
    "simes_df = pd.DataFrame(alt_values, columns = [\"gene\", \"simes_p\", \"harmonic_p\"])\n",
    "m = len(simes_df)\n",
    "\n",
    "print(\"------\")\n",
    "\n",
    "signif = set()\n",
    "alpha = 0.1\n",
    "simes_df = simes_df.sort_values(by = \"simes_p\")\n",
    "for counter, (index, row) in enumerate(simes_df.iterrows()):\n",
    "    if row[\"simes_p\"] <= ((counter+1) / m) * alpha:\n",
    "        signif.add(row[\"gene\"])\n",
    " \n",
    "\n",
    "\n",
    "print(len(signif))\n",
    "print(\", \".join(list(signif)))\n",
    "print(\"less than 0.05\", len(simes_df.loc[simes_df[\"simes_p\"] <= 0.05]))\n",
    "\n",
    "title = \"Deleteriousness burden, genome-wide\"\n",
    "all_vals.append([title, m, len(signif), \"|\".join(sorted(list(signif)))])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a18ad0",
   "metadata": {},
   "source": [
    "### Regeneron burden, genome-wide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "556d9b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"./data/GCST90083019_buildGRCh38.tsv\", delimiter = \"\\t\")\n",
    "df[\"gene\"] = df[\"Name\"].apply(lambda x : x.split(\"(\")[0])\n",
    "df = df.loc[\n",
    "    (~df[\"p_value\"].isna())\n",
    "]\n",
    "alt_values = []\n",
    "for g in set(df[\"gene\"].values):\n",
    "    s = df.loc[df[\"gene\"] == g]\n",
    "    p_vals = pd.Series(s[\"p_value\"].values)\n",
    "    simes_p = simes(p_vals)\n",
    "    harmonic_p = get_harmoic_p(p_vals.values)\n",
    "    alt_values.append([g, simes_p, harmonic_p])\n",
    "    \n",
    "simes_df = pd.DataFrame(alt_values, columns = [\"gene\", \"simes_p\", \"harmonic_p\"])\n",
    "\n",
    "signif = set()\n",
    "alpha = 0.1\n",
    "simes_df = simes_df.sort_values(by = \"simes_p\")\n",
    "m = len(simes_df)\n",
    "\n",
    "for counter, (index, row) in enumerate(simes_df.iterrows()):\n",
    "    if row[\"simes_p\"] <= ((counter+1) / m) * alpha:\n",
    "        signif.add(row[\"gene\"])\n",
    "    \n",
    "    \n",
    "print(len(signif), m - len(signif))\n",
    "print(\", \".join(sorted(list(signif))))\n",
    "print(\"less than 0.05\", len(simes_df.loc[simes_df[\"simes_p\"] <= 0.05]))\n",
    "\n",
    "title = \"Regeneron burden, genome-wide\"\n",
    "all_vals.append([title, m, len(signif), \"|\".join(sorted(list(signif)))])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2832ea0f",
   "metadata": {},
   "source": [
    "### Deleteriousness burden, module 255 (up to 14 genes, may be fewer depending on which have variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb9ae6ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "mega_df = pd.concat([\n",
    "    pd.read_csv('041522_whole_genome_0th_perc.csv'),\n",
    "    pd.read_csv(f\"041522_whole_genome_25th_perc.csv\"),\n",
    "    pd.read_csv(f\"041522_whole_genome_50th_perc.csv\"),\n",
    "    pd.read_csv(f\"041522_whole_genome_75th_perc.csv\")\n",
    "])\n",
    "\n",
    "df = mega_df\n",
    "\n",
    "df[\"mwu_a_pval\"] = df[['mwu_a_pval_greater', 'mwu_a_pval_less']].min(axis=1)\n",
    "df[\"mwu_two_sided\"] = df[\"mwu_a_pval\"].apply(lambda x : x*2)\n",
    "df[\"mwu_a_pval_intra\"] = df[['mwu_a_pval_intra_greater','mwu_a_pval_intra_less']].min(axis=1)\n",
    "\n",
    "col_of_interest = \"mwu_a_pval\"\n",
    "\n",
    "subset = df.loc[\n",
    "    (df[\"gene\"].isin(module_255_genes)) &\n",
    "    (~df[col_of_interest].isna()) &\n",
    "    (df['high_impact_carriers'] >= 10)\n",
    "].sort_values(by = col_of_interest)\n",
    "\n",
    "alt_values = []\n",
    "for g in set(subset[\"gene\"].values):\n",
    "    s = subset.loc[subset[\"gene\"] == g]\n",
    "    p_vals = pd.Series(s[col_of_interest].values)\n",
    "    simes_p = simes(p_vals)\n",
    "    harmonic_p = get_harmoic_p(p_vals.values)\n",
    "    alt_values.append([g, simes_p, harmonic_p])\n",
    "    \n",
    "simes_df = pd.DataFrame(alt_values, columns = [\"gene\", \"simes_p\", \"harmonic_p\"])\n",
    "m = len(simes_df)\n",
    "print(m)\n",
    "print(\"------\")\n",
    "\n",
    "signif = set()\n",
    "alpha = 0.1\n",
    "simes_df = simes_df.sort_values(by = \"simes_p\")\n",
    "for counter, (index, row) in enumerate(simes_df.iterrows()):\n",
    "    if row[\"simes_p\"] <= ((counter+1) / m) * alpha:\n",
    "        signif.add(row[\"gene\"])\n",
    "    \n",
    "print(len(signif))\n",
    "print(\", \".join(list(signif)))\n",
    "\n",
    "\n",
    "print(\"less than 0.05\", len(simes_df.loc[simes_df[\"simes_p\"] <= 0.05]))\n",
    "\n",
    "title = \"Deleteriousness burden, module 255 (up to 14 genes, may be fewer depending on which have variants)\"\n",
    "all_vals.append([title, m, len(signif), \"|\".join(sorted(list(signif)))])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd8dd061",
   "metadata": {},
   "source": [
    "### Regeneron burden, module 255 (up to 14 genes, may be fewer depending on which have variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5bb9129",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"./data/GCST90083019_buildGRCh38.tsv\", delimiter = \"\\t\")\n",
    "df[\"gene\"] = df[\"Name\"].apply(lambda x : x.split(\"(\")[0])\n",
    "df = df.loc[\n",
    "    (df[\"gene\"].isin(module_255_genes)) &\n",
    "    (~df[\"p_value\"].isna())\n",
    "]\n",
    "alt_values = []\n",
    "for g in set(df[\"gene\"].values):\n",
    "    s = df.loc[df[\"gene\"] == g]\n",
    "    p_vals = pd.Series(s[\"p_value\"].values)\n",
    "    simes_p = simes(p_vals)\n",
    "    harmonic_p = get_harmoic_p(p_vals.values)\n",
    "    alt_values.append([g, simes_p, harmonic_p])\n",
    "    \n",
    "simes_df = pd.DataFrame(alt_values, columns = [\"gene\", \"simes_p\", \"harmonic_p\"])\n",
    "\n",
    "signif = set()\n",
    "alpha = 0.1\n",
    "simes_df = simes_df.sort_values(by = \"simes_p\")\n",
    "m = len(simes_df)\n",
    "\n",
    "for counter, (index, row) in enumerate(simes_df.iterrows()):\n",
    "    if row[\"simes_p\"] <= ((counter+1) / m) * alpha:\n",
    "        signif.add(row[\"gene\"])\n",
    "    \n",
    "print(len(signif), m - len(signif))\n",
    "print(\", \".join(sorted(list(signif))))\n",
    "\n",
    "print(\"less than 0.05\", len(simes_df.loc[simes_df[\"simes_p\"] <= 0.05]))\n",
    "\n",
    "title = \"Regeneron burden, module 255 (up to 14 genes, may be fewer depending on which have variants)\"\n",
    "all_vals.append([title, m, len(signif), \"|\".join(sorted(list(signif)))])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "037c7835",
   "metadata": {},
   "source": [
    "### Deleteriousness burden, module 69 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8359be92",
   "metadata": {},
   "outputs": [],
   "source": [
    "mega_df = pd.concat([\n",
    "    pd.read_csv('041522_whole_genome_0th_perc.csv'),\n",
    "    pd.read_csv(f\"041522_whole_genome_25th_perc.csv\"),\n",
    "    pd.read_csv(f\"041522_whole_genome_50th_perc.csv\"),\n",
    "    pd.read_csv(f\"041522_whole_genome_75th_perc.csv\")\n",
    "])\n",
    "\n",
    "df = mega_df\n",
    "\n",
    "df[\"mwu_a_pval\"] = df[['mwu_a_pval_greater', 'mwu_a_pval_less']].min(axis=1)\n",
    "df[\"mwu_two_sided\"] = df[\"mwu_a_pval\"].apply(lambda x : x*2)\n",
    "df[\"mwu_a_pval_intra\"] = df[['mwu_a_pval_intra_greater','mwu_a_pval_intra_less']].min(axis=1)\n",
    "\n",
    "\n",
    "col_of_interest = \"mwu_a_pval\"\n",
    "\n",
    "subset = df.loc[\n",
    "    (df[\"gene\"].isin(module_69_genes)) &\n",
    "    (~df[col_of_interest].isna()) &\n",
    "    (df['high_impact_carriers'] >= 10)\n",
    "].sort_values(by = col_of_interest)\n",
    "\n",
    "alt_values = []\n",
    "for g in set(subset[\"gene\"].values):\n",
    "    s = subset.loc[subset[\"gene\"] == g]\n",
    "    p_vals = pd.Series(s[col_of_interest].values)\n",
    "    simes_p = simes(p_vals)\n",
    "    harmonic_p = get_harmoic_p(p_vals.values)\n",
    "    alt_values.append([g, simes_p, harmonic_p])\n",
    "    \n",
    "simes_df = pd.DataFrame(alt_values, columns = [\"gene\", \"simes_p\", \"harmonic_p\"])\n",
    "m = len(simes_df)\n",
    "print(m)\n",
    "print(\"------\")\n",
    "\n",
    "signif = set()\n",
    "alpha = 0.1\n",
    "simes_df = simes_df.sort_values(by = \"simes_p\")\n",
    "for counter, (index, row) in enumerate(simes_df.iterrows()):\n",
    "    if row[\"simes_p\"] <= ((counter+1) / m) * alpha:\n",
    "        signif.add(row[\"gene\"])\n",
    "    \n",
    "print(len(signif))\n",
    "print(\", \".join(list(signif)))\n",
    "\n",
    "print(\"less than 0.05\", len(simes_df.loc[simes_df[\"simes_p\"] <= 0.05]))\n",
    "\n",
    "\n",
    "title = \"Deleteriousness burden, module 69\"\n",
    "all_vals.append([title, m, len(signif), \"|\".join(sorted(list(signif)))])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65337062",
   "metadata": {},
   "source": [
    "### Regeneron burden, module 69 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f55ee0ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"./data/GCST90083019_buildGRCh38.tsv\", delimiter = \"\\t\")\n",
    "df[\"gene\"] = df[\"Name\"].apply(lambda x : x.split(\"(\")[0])\n",
    "df = df.loc[\n",
    "    (df[\"gene\"].isin(module_69_genes)) &\n",
    "    (~df[\"p_value\"].isna())\n",
    "]\n",
    "alt_values = []\n",
    "for g in set(df[\"gene\"].values):\n",
    "    s = df.loc[df[\"gene\"] == g]\n",
    "    p_vals = pd.Series(s[\"p_value\"].values)\n",
    "    simes_p = simes(p_vals)\n",
    "    harmonic_p = get_harmoic_p(p_vals.values)\n",
    "    alt_values.append([g, simes_p, harmonic_p])\n",
    "    \n",
    "simes_df = pd.DataFrame(alt_values, columns = [\"gene\", \"simes_p\", \"harmonic_p\"])\n",
    "\n",
    "signif = set()\n",
    "alpha = 0.1\n",
    "simes_df = simes_df.sort_values(by = \"simes_p\")\n",
    "m = len(simes_df)\n",
    "\n",
    "for counter, (index, row) in enumerate(simes_df.iterrows()):\n",
    "    if row[\"simes_p\"] <= ((counter+1) / m) * alpha:\n",
    "        signif.add(row[\"gene\"])\n",
    "    \n",
    "print(len(signif), m - len(signif))\n",
    "print(\", \".join(sorted(list(signif))))\n",
    "print(\"less than 0.05\", len(simes_df.loc[simes_df[\"simes_p\"] <= 0.05]))\n",
    "\n",
    "title = \"Regeneron burden, module 69\"\n",
    "all_vals.append([title, m, len(signif), \"|\".join(sorted(list(signif)))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0800c6d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(all_vals, columns = [\"comparison\", \"n genes\", \"n signif\", \"significant genes\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9416f5e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"VEST4_Regeneron_significant_comparisons_8_conditions.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
